src/Pulsar.Client/Auth/AuthenticationOauth2.fs

```diff
@@ -0,0 +1,115 @@
+﻿module Pulsar.Client.Auth.Oauth2
+
+open System
+open System.IO
+open System.Text.Json
+open System.Net.Http
+open System.Text.Json.Serialization
+open Pulsar.Client.Api
+open Pulsar.Client.Auth.Oauth2Token
+open FSharp.Control.Tasks.V2.ContextInsensitive
+type Metadata =
+   {
+        [<JsonPropertyName("issuer")>]
+        Issuer:string
+        
+        [<JsonPropertyName("token_endpoint")>]
+        TokenEndpoint:string
+        
+        [<JsonPropertyName("userinfo_endpoint")>]
+        UserInfoEndpoint:string
+        
+        [<JsonPropertyName("revocation_endpoint")>]
+        RevocationEndpoint:string  
+               
+        [<JsonPropertyName("jwks_uri")>]
+        JwksUri:string
+        
+        [<JsonPropertyName("device_authorization_endpoint")>]
+        DeviceAuthorizationEndpoint:string 
+    }
+type Credentials =
+    {  
+        [<JsonPropertyName("type")>]
+        CredsType : string
+        
+        [<JsonPropertyName("client_id")>]
+        ClientId : string
+        
+        [<JsonPropertyName("client_secret")>]
+        ClientSecret : string 
+
+        [<JsonPropertyName("client_email")>]
+        ClientEmail : string 
+
+        [<JsonPropertyName("issuer_url")>]
+        IssuerUrl : string 
+    }
+
+//Gets a well-known metadata URL for the given OAuth issuer URL.
+//https://tools.ietf.org/id/draft-ietf-oauth-discovery-08.html#ASConfig
+let getWellKnownMetadataUrl (issuerUrl: Uri) : Uri =
+    Uri(issuerUrl.AbsoluteUri + ".well-known/openid-configuration")
+
+let getMetadata (issuerUrl: Uri)  =
+    task {
+        use client = new HttpClient()        
+        let metadataDataUrl = getWellKnownMetadataUrl issuerUrl        
+        let! response = client.GetStreamAsync metadataDataUrl 
+        return! JsonSerializer.DeserializeAsync<Metadata> response        
+    }
+let createClient issuerUrl =
+    task {
+        let! data = getMetadata issuerUrl 
+        return TokenClient(Uri(data.TokenEndpoint))
+    }
+    
+let openAndDeserializeCreds uri =
+    task{
+        use fs = new FileStream(uri, FileMode.Open)
+        let! temp = JsonSerializer.DeserializeAsync<Credentials>(fs)
+        return temp
+    }
+    
+type AuthenticationOauth2(issuerUrl: Uri, credentials: Uri, audience: Uri) =
```

Comment:
Audience is not Uri in Java version, can you please also rename credentials to privateKey (like in Java version)